### PR TITLE
Pmp 143

### PR DESCRIPTION
**Pull request by jeanericdu31**
Must wait at least one valid review before merge.

### CODE REVIEW

File: iec104.cpp [hunk 1]
```diff
@@ -18,6 +18,20 @@
 #include <iostream>
 #include <utility>
 
+// Qualifier of command
+#define NoAddDefinition 0
```
Comment (jeanericdu31) [inline iec104.cpp:22]: Why didn't you use static variables instead of #define?

**Review by jeanericdu31 — COMMENTED** [review]
You should use static const members instead of #define.

File: iec104.cpp [hunk 2]
```diff
@@ -963,11 +1033,21 @@ void IEC104Client::m_addData(vector<Datapoint*>& datapoints, int64_t ioa,
         if (feature.value() == "ioa")
             measure_features->push_back(m_createDatapoint(feature.key(), ioa));
         else if (feature.value() == "value")
-            measure_features->push_back(
-                m_createDatapoint(feature.key(), value));
+        {
+            if (typeid(valuestate).name() == typeid(bool).name())
```
Comment (ArnauldAlex) [inline iec104.cpp:1037]: this new code is not covered by branch, at least I don't see in code the test of the condition


https://sonarcloud.io/component_measures?metric=new_coverage&selected=fledge-power_fledge-south-iec104%3Aiec104.cpp&view=list&pullRequest=10&id=fledge-power_fledge-south-iec104

File: iec104.cpp (hunk 1 above)
Comment (ArnauldAlex) [reply iec104.cpp:22]: I agree, why not using simple class var ?

File: iec104.cpp [hunk 3]
```diff
@@ -182,10 +197,12 @@ bool IEC104::m_asduReceivedHandler(void* parameter, int address,
             Logger::getLogger()->info("Test command with time tag CP56Time2a");
             break;
         case C_SC_TA_1:
+            handleASDU(labels, datapoints, mclient, asdu, handleC_SC_TA_1);
             Logger::getLogger()->info(
```
Comment (ArnauldAlex) [inline iec104.cpp:201]: As the other case, as it's now handled I think you can remove the logs

File: iec104.cpp [hunk 4]
```diff
@@ -182,10 +197,12 @@ bool IEC104::m_asduReceivedHandler(void* parameter, int address,
...
                 "Single command with time tag CP56Time2a");
             break;
         case C_DC_TA_1:
+            handleASDU(labels, datapoints, mclient, asdu, handleC_DC_TA_1);
             Logger::getLogger()->info(
```
Comment (ArnauldAlex) [inline iec104.cpp:206]: same remark as above

File: iec104.cpp [hunk 5]
```diff
@@ -463,6 +480,59 @@ void IEC104::handleM_ME_TF_1(vector<Datapoint*>& datapoints, string& label,
...
+// Commands and Set-Point Commands
+// SingleCommandWithCP56Time2a
+void IEC104::handleC_SC_TA_1(vector<Datapoint*>& datapoints, string& label,
+                             IEC104Client* mclient, unsigned int& ca,
+                             CS101_ASDU& asdu, InformationObject& io,
+                             uint64_t& ioa)
+{
+    auto io_casted = (SingleCommandWithCP56Time2a)io;
+    bool state = SingleCommand_getState((SingleCommand)io_casted);
+    // qu qualifier of command
```
Comment (ArnauldAlex) [inline iec104.cpp:492]: this comment is not useful

File: iec104.cpp [hunk 6]
```diff
@@ -463,6 +480,59 @@ void IEC104::handleM_ME_TF_1(vector<Datapoint*>& datapoints, string& label,
...
+
+// DoubleCommandWithCP56Time2a
+void IEC104::handleC_DC_TA_1(vector<Datapoint*>& datapoints, string& label,
+                             IEC104Client* mclient, unsigned int& ca,
+                             CS101_ASDU& asdu, InformationObject& io,
+                             uint64_t& ioa)
+{
+    auto io_casted = (DoubleCommandWithCP56Time2a)io;
+    int state = DoubleCommand_getState((DoubleCommand)io_casted);
+    // qu qualifier of command
```
Comment (ArnauldAlex) [inline iec104.cpp:517]: same as above

File: iec104.cpp (hunk 2 above)
Comment (ArnauldAlex) [inline iec104.cpp:1037]: you can remove ".name" to compare two type id you don't need this information.

File: include/iec104.h [hunk 7]
```diff
@@ -229,7 +240,7 @@ class IEC104Client
     template <class T>
     void m_addData(std::vector<Datapoint*>& datapoints, int64_t ioa,
                    const std::string& dataname, const T value,
```
Comment (ArnauldAlex) [inline include/iec104.h:242]: for consistency rename (or revert the rename in .cpp) value and qd (also in public declaration)

File: iec104.cpp (hunk 2 above)
Comment (ArnauldAlex) [inline iec104.cpp:1037]: m_addData is a private function, that is called by two public function. In each; value take the type int64_t or float. 
So why do we test if the type of value is bool ?

**Review by ArnauldAlex — CHANGES_REQUESTED** [review]

**Review by ArnauldAlex — APPROVED** [review]

**Comment by ArnauldAlex** [comment]
C'est ok pour la review @aklira @jeanericdu31

**Pull request merged**